### PR TITLE
[spec] Add security and privacy considerations sections

### LINKED ISSUE #29: Fill in security and privacy considerations
As mentioned in the TODO at https://wicg.github.io/scheduling-apis/#sec-security.

### PR DESCRIPTION
**Pull request by shaseley**
Closes #29 

@jyasskin can you PTAL? Thanks!

***
<a href="https://pr-preview.s3.amazonaws.com/shaseley/main-thread-scheduling/pull/34.html" title="Last updated on Jun 9, 2021, 11:06 PM UTC (ecd2663)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/scheduling-apis/34/5df04f7...shaseley:ecd2663.html" title="Last updated on Jun 9, 2021, 11:06 PM UTC (ecd2663)">Diff</a>

### CODE REVIEW

File: spec/security.md [hunk 1]
```diff
@@ -1,4 +1,64 @@
...
+<div class="non-normative">
+The first consideration is whether the {{Scheduler/postTask()}} API can be used
+as a high-resolution timing source. {{Scheduler/postTask()}}'s
+{{SchedulerPostTaskOptions/delay}}, like
+{{WindowOrWorkerGlobalScope/setTimeout()}}'s timeout value, is expressed in
+whole milliseconds (the minimum non-zero delay being 1 ms), and there is no
+guarantee that tasks will run exactly when the delay expires since tasks are
+queued when the delay expires. Given this, we do not believe the API can be
+used as a high-resolution timing source, but we mention it here because of the
+general interest in the topic.
```
Comment (jyasskin) [inline spec/security.md:24]:
```suggestion
used as a high-resolution timing source.
```

I don't think the last clause adds anything for readers.

File: spec/security.md [hunk 2]
```diff
@@ -1,4 +1,64 @@
...
+
+<div class="non-normative">
+The first consideration is whether the {{Scheduler/postTask()}} API can be used
+as a high-resolution timing source. {{Scheduler/postTask()}}'s
+{{SchedulerPostTaskOptions/delay}}, like
+{{WindowOrWorkerGlobalScope/setTimeout()}}'s timeout value, is expressed in
+whole milliseconds (the minimum non-zero delay being 1 ms), and there is no
+guarantee that tasks will run exactly when the delay expires since tasks are
+queued when the delay expires. Given this, we do not believe the API can be
+used as a high-resolution timing source, but we mention it here because of the
```
Comment (jyasskin) [inline spec/security.md:23]: Try to lead with the lede. So something like,

> This API cannot be used as a high-resolution timing source. Like setTimeout's timeout value, postTask()'s delay is expressed in whole milliseconds (the minimum non-zero delay is 1 ms), so callers can't express any timing more precise than 1ms. Further, the fact that tasks are only queued when their delay expires, and not run instantly, further reduces the precision available to callers.

That also gives future API tweakers a clue to look closely at attempts to change those two qualities.

File: spec/security.md [hunk 3]
```diff
@@ -1,4 +1,64 @@
...
+monitoring when their tasks run. For example, an attacker could flood the
+system with {{TaskPriority/user-blocking}} tasks and expect them to run
+consecutively; if there are large gaps in between, then the attacker might
+infer something about tasks running in another event loop, e.g. they have
+higher priority.
+
+The first thing we note is that the attacker would not be able to definitively
+tell that a task ran in another event loop, e.g. an internal browser task might
+have run, or perhaps the UA throttled the attacker's script.  Second, any
+information gained would be implementation-dependent since inter-event-loop
```
Comment (jyasskin) [inline spec/security.md:46]: I'm not sure it's really a defense that the attacker would have to implement their attack for each of ~3 implementations.

File: spec/security.md [hunk 4]
```diff
@@ -1,4 +1,64 @@
...
+general interest in the topic.
+</div>
+
+Monitoring Another Origin's Tasks {#sec-security-monitoring-tasks}
+---------------------
+
+<div class="non-normative">
+The second consideration is whether {{Scheduler/postTask()}} leaks any
+information about other origins' tasks. The threat model we consider is code
+from two different origins running in separate event loops in the same thread.
```
Comment (jyasskin) [inline spec/security.md:33]: It's not quite right to call this the "threat model". Maybe just "We consider an attacker running on one origin trying to obtain information about code executing in another origin (and hence in a separate event loop) that is scheduled in the same thread in a browser."

File: spec/security.md [hunk 5]
```diff
@@ -1,4 +1,64 @@
...
+
+Because the UA can only run tasks from one event loop at a time, an attacker
```
Comment (jyasskin) [inline spec/security.md:35]: "a thread within the UA"?

File: spec/security.md [hunk 6]
```diff
@@ -1,4 +1,64 @@
...
+might be able to gain information about tasks running in another event loop by
+monitoring when their tasks run. For example, an attacker could flood the
+system with {{TaskPriority/user-blocking}} tasks and expect them to run
+consecutively; if there are large gaps in between, then the attacker might
+infer something about tasks running in another event loop, e.g. they have
+higher priority.
```
Comment (jyasskin) [inline spec/security.md:41]: What does it mean for a task to have higher priority than a user-blocking one? Is it just that the attacker learns that the other colocated origin had some work to do, and the UA decided to give it a chance to run?

File: spec/security.md [hunk 7]
```diff
@@ -1,4 +1,64 @@
...
+
+The first thing we note is that the attacker would not be able to definitively
+tell that a task ran in another event loop, e.g. an internal browser task might
+have run, or perhaps the UA throttled the attacker's script.  Second, any
+information gained would be implementation-dependent since inter-event-loop
+task selection is not specified. Our opinion is that information gained in such
+an attack is likely to be benign, but there are mitigations that implementers
```
Comment (jyasskin) [inline spec/security.md:48]: Have any of the xsleaks folks looked at this? They've been pretty clever about making leaks non-benign.

File: spec/security.md [hunk 8]
```diff
@@ -1,4 +1,64 @@
...
+can consider to minimize the risk:
```
Comment (jyasskin) [inline spec/security.md:49]:
```suggestion
an attack is likely to be benign, but implementers can use the following mitigations to minimize the risk:
```

Is Chrome implementing these? e.g. I think all the site isolation work and out-of-process iframes implement the first.

File: spec/security.md (hunk 6 above)
Comment (shaseley) [reply spec/security.md:41]: That's what I was intending it to mean. I'll try to rewrite this so it's clearer. The problem is, as I mention below, whatever information is gained is implementation-dependent, so anything here will be hypothetical. And realistically, the information gained is likely to be probabilistic. I'm not sure the best way to express all of that in this section.

**Review by shaseley — COMMENTED** [review]
@jyasskin I'll work on rewriting this and uploading a new version, but I wanted to reply to your comments first in case you have thoughts.

File: spec/security.md (hunk 3 above)
Comment (jyasskin) [reply spec/security.md:46]: Maybe "The information exposed would depend on implementation details, and implementations can reduce it as described below."

File: spec/security.md (hunk 8 above)
Comment (jyasskin) [reply spec/security.md:49]: \o/

File: spec/security.md (hunk 7 above)
Comment (shaseley) [reply spec/security.md]: I removed this statement and rewrote this section, attempting to explain what the status quo is and what the delta we're introducing is --- without making any kind of assertions of the severity.

**Comment by shaseley** [comment]
Hi @jyasskin, can you PTAL? Thanks!

File: explainers/security-privacy-questionnaire-post-task.md [hunk 9]
```diff
@@ -131,6 +132,9 @@ They work the same as in the non-private mode.
 > 16. Does this specification have both "Security Considerations" and "Privacy
 >     Considerations" sections?
 
+Yes ([security](https://wicg.github.io/scheduling-apis/#sec-security),
+[privacy](https://wicg.github.io/scheduling-apis/#sec-privacy)).
+
 [Not yet](https://github.com/WICG/scheduling-apis/issues/29).
```
Comment (jyasskin) [inline explainers/security-privacy-questionnaire-post-task.md:138]: Remove this line now?

**Review by jyasskin — APPROVED** [review]
Thanks!

File: explainers/security-privacy-questionnaire-post-task.md (hunk 9 above)
Comment (shaseley) [reply explainers/security-privacy-questionnaire-post-task.md:138]: Ah, good catch! Will do.

**Pull request merged**